src/main/java/org/truffleruby/core/array/ArrayConcatNode.java

```diff
@@ -35,6 +36,10 @@ public ArrayConcatNode(RubyNode[] children) {
 
     @Override
     public DynamicObject execute(VirtualFrame frame) {
+        if (arrayBuilderNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            arrayBuilderNode = insert(ArrayBuilderNode.create());
+        }
```

Comment:
It gets a bit tiring writing this pattern out doesn't it? I wonder if the DSL could do something? Maybe we should never be writing nodes that aren't DSL nodes, so we can always use `@Cached`? Maybe we could have a DSL light, where you specify an `abstract ArrayBuilderNode getArrayBuilderNode()` and the DSL fills that in for you.

Reply:
Yes, agree to everything. Having some special support for that would be great. Using the DSL and @Cached makes it easier, but only if the node is used just by this one specialization. If the DSL could generate such getters automatically, that would be great (even though it would violate your pattern of including the execute method).

In the end, it's writing 4 lines (6, if in a function) in ~100 places. 400 LOC versus 25 MB of footprint, I think that pays off even without further support.